### PR TITLE
Refine linkToVirtual to a direct call when no vtable entry exists

### PR DESCRIPTION
**Pull request by nbhuiyan**
MethodHandleTransformer refines linkToVirtual calls to the target
method when MemberName is known. The call kind of the refined
linkToVirtual call target was always treated as Virtual.
However, for private virtual methods and java/lang/Object;
type virtual methods, there is no corresponding entry in the vtable,
and for such methods the interpreter vtable index is set as 0.
The dispatch is not performed through the vtable entry, but instead
it is a direct call to the target method obtained from MemberName.

By checking if jitVtableOffset is positive, we can determine if
a call needs to be dispatched through the vTable. Otherwise, we
can handle it similar to linkToStatic.

Closes: #13430 

Signed-off-by: Nazim Bhuiyan <nubhuiyan@ibm.com>

### CODE REVIEW

**Comment by nbhuiyan** [comment]
@0xdaryl @knn-k @Akira1Saitoh FYI

**Comment by nbhuiyan** [comment]
@0xdaryl requesting review

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk17

Getting a head start on the testing...

**Comment by 0xdaryl** [comment]
Jenkins test sanity alinux64 jdk17

File: runtime/compiler/optimizer/J9TransformUtil.cpp [hunk 1]
```diff
@@ -2180,26 +2180,33 @@ J9::TransformUtil::refineMethodHandleLinkTo(TR::Compilation* comp, TR::TreeTop*
...
-      }
+      vTableSlot = fej9->vTableOrITableIndexFromMemberName(comp, mnIndex);
+      jitVTableOffset = fej9->vTableSlotToVirtualCallOffset(vTableSlot);
+      // For private virtual methods and java/lang/Object; type virtual methods, there is no corresponding
+      // entry in the vtable, and for such methods the interpreter vtable index is 0.
+      // The dispatch is not performed through the vtable entry, but directly dispatched to the J9Method
+      // in Membername.vmtarget. In such cases, we can refine a linkToVirtual call similar to linkToStatic.
+      // jitVTableOffset is obtained using sizeof(J9Class) - interpreter vtable offset, resulting in a positive
+      // value when the interpreter vtable index we get (from MemberName.vmindex.vtableoffset) is set as 0.
+      if (jitVTableOffset > 0)
```
Comment (0xdaryl) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:2195]: I'm confused by this test.  If the problematic interpreter vtable offset is "sizeof(J9Class)" or +376 then shouldn't this test be `jitVTableOffset >= 0` ?

Comment (nbhuiyan) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:2195]: The problematic interpreter vtable offset is when it is set as `0`, resulting in the jitVtableOffset being `sizeof(J9Class)`. Normally, interpreter vtable offset is set a value `> sizeof(J9Class)` so the vtable offset we obtain will always be negative, which is why I thought a simple test for a positive integer value was sufficient. We can't have `jitVtableOffset == 0` as that is the start of the J9Class. Do you think changing the operator to `>=` would provide more clarity?

Comment (nbhuiyan) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:2195]: Just for reference, here is where the interpreter vtable offset is set:
https://github.com/eclipse-openj9/openj9/blob/7ad86703975e3432b6ec90b998ab04aa1a835e71/runtime/vm/createramclass.cpp#L1407-L1421

`J9VTABLE_OFFSET_FROM_INDEX` takes the index in the vtable and adds `sizeof(J9Class) + sizeof(J9VTableHeader)`, and the vtable entries begin following the J9VTableHeader. So whenever there's a vtable entry, the jitVTableOffset will end up being a negative value.

**Comment by 0xdaryl** [comment]
Jenkins test sanity alinux64 jdk17

**Comment by 0xdaryl** [comment]
Internal AArch64 testing succeeded (finally--after some infra problems).

**Review by 0xdaryl — APPROVED** [review]

**Comment by 0xdaryl** [comment]
Please create a PR for the 0.28.0 branch.

**Pull request merged**